### ISSUE TITLE
Wrong graph optimization with dynamic shape

### ISSUE THREAD

**Issue by wodesuck** [microsoft/onnxscript#2577]
```python
import torch
import torch.nn as nn


class Model(nn.Module):
    def forward(self, x):
        x[:2] = 0
        return x


def main():
    model = Model()
    args = (torch.rand(10),)
    dim = torch.export.Dim("dim")
    dynamic_shapes = {"x": {0: dim}}
    torch.onnx.export(
        model,
        args,
        "model_test.onnx",
        dynamic_shapes=dynamic_shapes,
        dynamo=True,
    )


if __name__ == '__main__':
    main()
```

The output graph will be optimized to:

<img width="282" height="290" alt="Image" src="https://github.com/user-attachments/assets/1ba5cfb9-a652-4d39-b708-43cd44e6d56d" />

Graph with `optimize=False` will be:

<img width="2200" height="1110" alt="Image" src="https://github.com/user-attachments/assets/14adca28-dd95-4a57-a824-f1c2b4956558" />

And This is what happen: [collapse_slice2](https://github.com/microsoft/onnxscript/blob/main/onnxscript/rewriter/rules/common/_collapse_slices.py#L98) found that input shape and output shape of Slice (the right one in the graph) are "same" (they are both `[None]`), so it is optmized to Identity. Then it match the rule [ScatterAllDynamic](https://github.com/microsoft/onnxscript/blob/main/onnxscript/rewriter/rules/common/_redundant_scatter_nd.py#L27), the whole graph is optimized to an Identity op.

The root cause of trouble is [ir.SymbolicDim](https://github.com/onnx/ir-py/blob/main/src/onnx_ir/_core.py#L1112) allow None dim, and treat None == None. A simple solution to fix it is convert all None dim to a unique string, just need to add these two lines in `SymbolicDim.__init__`:

```python
 if value is None:
     value = "_s" + str(id(self))
```

**Comment by wodesuck** [comment]
@justinchuby I think this issue hasn't been completely resolved. #2583 only fixed the collapse slices rule, but there are other places where two None dimensions are incorrectly considered equal. Additionally, as I mentioned in [onnx/ir-py#199](https://github.com/onnx/ir-py/pull/199#issuecomment-3344884580), treating all None dimensions as unequal would also disable some good optimizations. I believe a better fix would be avoiding any None dimensions, when encountering an unknown dimension, we should always assign it a unique string identifier. This way, shape inference can recognize which dimensions are equal, while preventing unequal unknown dimensions from being accidentally considered equal.

**Comment by justinchuby** [comment]
Thanks - I think the None values are there because we don't have a shape inference engine that supports symbolic dimensions. All dynamic shapes from pytorch are properly recorded, but they may not be propagated correctly when we optimize because of our shape inference engine. Therefore setting strings on the None dims wouldn't help currently, I think?
